File: protocol-generator/php/src/Conn.php
<?php

declare(strict_types=1);

namespace EntityCore;

/**
 * Per-connection state carried through the §6.5 dispatch chain.
 *
 *   - handshake state (§4.1/§4.6): the issued nonce + the hello-declared peer_id;
 *   - `established` (post-authenticate gate);
 *   - the §6.13(b)/§6.11 reentry OUTBOUND seam: a callable that originates an
 *     EXECUTE back over THIS connection and returns the correlated response
 *     envelope (or null). In the single-thread event loop the callable pumps the
 *     loop until the reply correlates by request_id — there is no thread to block.
 *   - an outbound request_id counter (distinct from inbound request ids).
 */
final class Conn
{
    public bool $established = false;
    public ?string $issuedNonce = null;
    public ?string $helloPeerId = null;

    /** @var (callable(Envelope):?Envelope)|null */
    public $outbound = null;

    private int $outCounter = 0;

    public function nextOutCounter(): int
    {
        return ++$this->outCounter;
    }
}

File: protocol-generator/java/src/main/java/org/entitycore/protocol/peer/Conn.java
package org.entitycore.protocol.peer;

import java.util.function.Function;

/**
 * Per-connection state (§4.2 connection state is per-connection). Holds the §4.1
 * handshake progress (issued nonce, the initiator's claimed peer_id, established flag)
 * and the §6.13(b) handler-facing outbound seam.
 *
 * <p>The {@code outbound} seam sends an EXECUTE envelope over THIS connection and awaits
 * its correlated EXECUTE_RESPONSE (§6.11 reentry); the transport sets it. It is null
 * when the request did not arrive over a reentrant connection (e.g. an in-process call).
 */
public final class Conn {
    volatile boolean established;
    volatile byte[] issuedNonce;          // nonce we issued in our hello response
    volatile String helloPeerId;          // initiator's claimed peer_id from hello

    /** §6.13(b) reentry seam: send-and-await over this connection; null if unavailable. */
    volatile Function<Envelope, Envelope> outbound;
    private int outCounter;

    synchronized int nextOutCounter() {
        return ++outCounter;
    }
}
